numpy/_core/src/multiarray/conversion_utils.c

```diff
@@ -1352,3 +1353,24 @@ PyArray_IntTupleFromIntp(int len, npy_intp const *vals)
  fail:
     return intTuple;
 }
+
+/*
+ * Device string converter.
+ */
+NPY_NO_EXPORT int
+PyArray_DeviceConverterOptional(PyObject *object, NPY_DEVICE *device)
+{
+    if (object == Py_None) {
+        return NPY_SUCCEED;
+    }
+
+    if (PyUnicode_Check(object) &&
+        PyUnicode_Compare(object, npy_ma_str_cpu) == 0) {
+        *device = NPY_DEVICE_CPU;
+        return NPY_SUCCEED;
+    }
+
+    PyErr_SetString(PyExc_ValueError,
+            "Device not understood. Only \"cpu\" is allowed.");
```

Comment:
It's always helpful to add the received value in the error message, makes it way easier to understand what went wrong. So I suggest adding _", but received {device}"_ (needs to handle other strings, or repr of non-string input ideally)

Reply:
Sure! Added `", but received %S."` so `str(object)` is printed additionally in the error message.